### PR TITLE
Refactor Node Processing to allow Scene Multithreading

### PR DESCRIPTION
**Pull request by reduz**
* Node processing works on the concept of process groups.
* A node group can be inherited, run on main thread, or a sub-thread.
* Groups can be ordered.
* Process priority is now present for physics.

This is the first steps towards implementing https://github.com/godotengine/godot-proposals/issues/6424. No threading or thread guards exist yet.

Steps:

~~1. Ability to create thread groups in nodes.~~
~~2. Ability to run groups threaded.~~
~~3. Create message passing system.~~
~~4. Protect the whole Node API for thread misuse in debug mode.~~
~~5. Test and bechmark.~~

A small benchmark (near a hundred characters playing run animation in 3D) has a 3x speedup with this PR (it should have more, but after running vtune to investigate, there are many memory synchronization bottlenecks in the animation code that have to be fixed that are unrelated to this PR).

#### Setting up groups

Nodes now have a Thread Group option in the _process_ section. If a group is selected, this node and children nodes will now belong to a _thread group_. This means, that PROCESS and PHYSICS_PROCESS callbacks will be called in parallel for any node belonging to this thread group. Thread groups also have an _order_ variable. Groups are executed in-order from lesser to greater and everything in the same group is executed in parallel.

image

By default, all nodes belong to the _default_ group, which has process order 0, and processes in the main thread unless overriden with the `thread_group` option. This ensures backwards compatibility.

#### Synchronizing

Because it is common to want to use the results of the processing and set it to something that is not in the current scene, new functions were added to `Node`, such as : `call_deferred_thread_group`, which works like `call_deferred()`, but using the thread group.
With these functions, it is possible to send messages to a process group when its processed in a thread, as well as send messages from the process group to another one (be it in the main thread, or a sub-thread).

#### Thread safety

To ensure thread safety when processing groups, the following changes were made:

* SceneTree is now fully thread safe.
* Node is fully thread safe

The following error macros were added so they can be used in nodes (on the C++ side):

* `ERR_THREAD_GUARD` and `ERR_THREAD_GUARD_V` Use them at the beginning of functions to ensure that they are not being called from the wrong process group. 
* `ERR_MAIN_THREAD_GUARD` and `ERR_MAIN_THREAD_GUARD_V` Use them at the beginning of functions to ensure that this function can only be called from the main thread (calling them from groups will fail). Examples of functions that can only be called from the main thread are `Node::add_child` or `Node::remove_child`.

The functions above only perform checks if nodes are inside the SceneTree (so you can still build a scene in a thread, and then use `call_deferred()` to put it in the main thread). Additionally, the checks above only take place on _debug builds_. They are not compiled during _release builds_.

These functions also perform no locking or memory synchronization, they simply check the thread ID validity using TLS, so they 
are very efficient.

#### Testing

There is a new command-line option that allows running Godot while forcing single-threaded mode in order to compare or debug:
```
$ godot --single-threaded-scene
```

## How do you normally use this?

The idea is that, in Godot, most scenes are generally self contained units. An enemy, a bullet, an NPC, etc. If those scenes have something that does significant complex processing (such as animation playback, character physics, pathfinding, etc), then these scenes will greatly benefit from running their `process` iteration in parallel when there is a significant amount of them (dozens or hundreds).

If they do something simple (like, just motion), then its likely there will be not much of a performance improvement and the performance could be affected instead. For those cases (lots of elements, simple logic), the [swarm proposal](https://github.com/godotengine/godot-proposals/issues/2380) is up as a proposed solution.

#### Protecting the Object API

Things that happen at Object level (set/get/set_meta/get_property_list/call/etc) are not protected. Those functions will need to become virtual and protections be added at Node level in the overrides.

#### Protecting the rest of the Node API

The Node API (all existing nodes) is quite large. Protecting all of it with the guard macros described above would ensure that the scene system is fully and completely thread-safe, but it's a lot of work to do it everywhere (and should be done as follow-up PRs). One possibility would do this only in the most common nodes for now (Control, Node2D, Node3D) and do the rest at a later time, gradually.

#### GDScript and other languages

GDScript should probably implement a call to ERR_THREAD_GUARD on functions inheriting from `Node` automatically (it should have no cost), which can only be disabled with some function annotation like @no_thread_guard or something.

For other languages, I think it should be possible to efficiently allow this logic manually by encapsulating a function call. Of course, it would be up to the user (if they don't do anything with thread groups, they don't need to do this).

#### Multiplayer API is not thread safe.

The multiplayer code (rpc as example) is not thread safe, the Node API does not care about this, so we need to discuss if we want to implement safety at Node level or Multiplayer level.

#### Some nodes need changes

* AnimationTreePlayer can't be run in parallel because the state variable is shared. This needs to be moved to thread local storage to enable running it in parallel.
* CharacterBody currently can't probably work in parallel because the physics server expects it to work in the main thread. This needs to be revised. It is possible it works, though.

#### Transform notifications are not threaded

Currently, accumulated transform notifications are not threaded, the list of notifications should be moved to the thread group. This would improve parallelism a bit more.

#### Synchronization pressure is not great

While for the most part the scene system runs well in single threaded mode, Godot abuses in many places of things that can seriously degrade multithreaded performance. Example of this is synchronization instructions (mutexes) and atomics (used a lot in types such as Vector, String, StringName, Ref<>, etc). The later can put serious memory pressure on multiple threads if the same memory is being accessed (as an example, a lot of nodes playing and accessing the same animation). This needs to be redesigned/fixed and proper benchmarks need to be done.

#### Multi threaded debugging

Merging this PR and making multi-threaded development easier means that the debugger ability to debug multiple threads has to finally be fixed.

#### Multi threaded profiling

A multithreaded profiler is overdue in Godot , so its possible to visualize all threads working in all frames over time.

## TODO (before merging this PR).

Agree on things, then relevant documentation can be filled.

### CODE REVIEW

**Comment by reduz** [comment]
Changed the approach and code, it should be a lot simpler now.

**Comment by fire** [comment]
There doesn't seem to be some errors in the cicd.

1. b/doc/classes/Node.xml changed
2. ERROR: ZIPReader must be opened before use in the sanitizer
3. Warning: scene\main\scene_tree.cpp(849): warning C4458: declaration of 'node_count' hides class member

**Comment by WolfgangSenff** [comment]
How (if at all) does this affect input functions across iOS and Android and other exports?

**Comment by reduz** [comment]
@WolfgangSenff Input still happens on the main thread, this is only for threading the process callbacks. Given most users just use the Input singleton, not much should change.

**Comment by fire** [comment]
Can you estimate the technical feasibility of a backport to Godot Engine 4.0?

**Comment by reduz** [comment]
@fire I think its very unlikely because this will be a very large change when completed.
But if you read [this article](https://godotengine.org/article/release-management-4-1/), means 4.1 will not be that far off.

**Comment by RevoluPowered** [comment]
@reduz do you think this will work with ENet by default to move it from the main thread to a separate thread?

Would be really handy for something I am working on.

File: editor/editor_node.cpp
Comment (m4gr3d) [inline editor/editor_node.cpp:636]: I think this should be an (advanced) editor setting so node threading for the editor can be toggled on/off.

Given the editor uses a wide surface of the engine apis, this would help with identifying bugs and issues with the node threading implementation. As an example, the Android port of the editor has already allowed us to identify bugs and issues with the engine on Android platforms.

Comment (m4gr3d) [reply editor/editor_node.cpp:636]: What do you mean by GUI will not support threading?

> Keep in mind that the way threading works at scene level

From the proposal and the screenshot above, it looks like threading can be configured _within a scene_ granularly at the node level. 
So just based off the screenshot, it would seem possible to configure different sections/nodes within the `EditorNode`  to run on different threads. I agree there may not be a reason to do so, but nonetheless the capability is there.

Comment (reduz) [reply editor/editor_node.cpp:636]: GUI will not support threading, because it does a lot of interactions at the tree level across nodes that can easily exit your thread workgroups and go somewhere else (like container resizing). The engine can catch this fine and throw an error, but ultimately if you have a large connected GUI you can't easily split it up in threads. It can be implemented at some point, but it would be a lot of work for very little benefit. Add to this that the GUI does very little amount of work in the process callbacks (which is what gets parallelized) so even if it could be parallelized the benefit would be very minimum.

Comment (m4gr3d) [reply editor/editor_node.cpp:636]: Thanks, that makes sense at the moment. 
Although it's worth pointing out that the use of Godot as a UI framework to develop applications is likely to increase over time, and most current UI frameworks also have the concept of splitting work across multiple threads, so that may be something worth considering in the future.

At the moment, yes that is not much of a concern.

File: doc/classes/Object.xml
Comment (adamscott) [inline doc/classes/Object.xml:360]: Add missing doc for `Object::cancel_free()`

Comment (adamscott) [reply doc/classes/Object.xml:360]: I would suggest something like:
```
Cancels the free of the object. 
[b]Note:[/b] Intended for internal engine use; calling this function directly will probably do nothing.
[b]Note:[/b] This is not by any means related to [method Node.queue_free].
```

**Comment by adamscott** [comment]
I created a test project, tell me @reduz if it is a appropriate benchmark.
[75901.zip](https://github.com/godotengine/godot/files/11434713/75901.zip)

Footage with each character in it's subthread
[Capture vidéo du 2023-05-09 14-06-17.webm](https://github.com/godotengine/godot/assets/270928/f5c3694c-3ea7-41db-8026-896f149db531)

Footage with `--single-threaded-scene`
[Capture vidéo du 2023-05-09 14-07-22.webm](https://github.com/godotengine/godot/assets/270928/f0e2ff98-4773-4fb2-b023-0b14772e8be5)

@Calinou I could use your benchmark tool in the test.

I used the Godot 3D Robot Character made by @arianagchow
https://captainripley.itch.io/godot-3d-robot-character
https://github.com/AGChow/3D-Godot-Robot-Platformer-Character

File: scene/main/node.cpp
Comment (adamscott) [inline scene/main/node.cpp:374]: Should this use `ERR_MAIN_THREAD_GUARD_V`?

**Comment by reduz** [comment]
@adamscott There is a good amount of things that need to be improved that are killing threaded performance right now in AnimationPlayer. After the PR is merged I will do a pass at fixing some things.

**Review by akien-mga — APPROVED** [review]
As discussed with other maintainers, this is a first step towards allowing proper scene multithreading, which is a work that will span multiple months and probably take until 4.2 or 4.3 to be fully stable. Until then, the feature will be flagged as experimental to make it clear that it's a work in progress, and might change from release to release.

This first step lays the foundation and shouldn't negatively impact the engine when the scene multithreading is not used, so we agreed to merge this now to unblock other parts of the work that will be based on this, even if it's not the fully finalized form that this will take.

@reduz also started working on some guidelines for contributors for writing thread safe scene code which will be shared soon.

**Comment by Targma** [comment]
I would appreciate if error regarding thread guard would include some identifiable information for node (name). Currently i have no good way to pinpoint errors working in C#.

**Pull request merged**
